### DISCUSSION TITLE
Metadata change not considered by Session as a change of a document

### DISCUSSION THREAD

**Discussion by mic2zar** [ravendb/ravendb#17784] [Q&A]
I have watched "Using RavenDB as a Queuing Infrastructure" from Oren and I am trying to (re)implement his idea for defer processing using @referesh flag in the metadata. 

What is not working for me is the update of @refresh flag. While it is getting saved when the entity is getting created (Command in the example), it is not getting refreshed when object already exists and is **attached** to the session.  How shall I communicate to the _dbSession that one of entities attached shall be resaved after metadata update. 

```
    public async Task AddOrUpdateAsync(Command command, DateTime refresh)
    {
        ArgumentNullException.ThrowIfNull(command);

        var existingCommand = await _dbSession.TryLoadCommandAsync(command);
        if (existingCommand != null)
        {   
           /// following line is not working ///
            _dbSession.Advanced.GetMetadataFor(existingCommand)["@refresh"] = refresh; 
        }
        else
        {
            await _dbSession.StoreAsync(command);
            _dbSession.Advanced.GetMetadataFor(command)["@refresh"] = refresh;
        }
    }
```

Even if I add SaveChangesAsync (which I should not because saving happens automatically for all our other objects on exist), the @refresh flag is not getting updated:

```
        if (existingCommand != null)
        {   
            _dbSession.Advanced.GetMetadataFor(existingCommand)["@refresh"] = refresh;
            await _dbSession.SaveChangesAsync();
        }
```

**Comment by georgiosd** [comment]
Did you enable refresh in the db?

Georgios

**Reply by mic2zar** [reply]
refresh feature is enabled :)

My question is about updating metadata property, which apparently as shown in the code in the case of @refresh flag can be set, but cannot be updated (dbsession does not consider it as a change that should be flashed back to database)

**Reply by georgiosd** [reply]
It can be updated, we have a lot of codel ike this production. The problem is somewhere else - I'd suggest attaching to a debugger :)

**Comment by mcrio** [comment]
Please can you make a failing test. 
It works for me. 

Did you perhaps disable  tracking session wide?

**Comment by mic2zar** [comment]
We investigated issue further and apparently `dbSession.Advanced.HasChanges` is not getting set whenever we update `metadata.@refresh` property. Is this intentional? We have in our code additional logic that was checking for this property. For the time being we disabled it, but that sounds like the bug (or at least it is highly unintuitive)  that dbSession.Advanced.HasChanges remains false after one of metadata properties has been updated.

**Reply by georgiosd** [reply]
Looks like you're right - I don't know why that is, maybe @mcrio can let us know.

Meanwhile this works:

![image](https://github.com/ravendb/ravendb/assets/1672054/8e4e53c7-15ec-4ed5-8dbd-418d941ad63b)

**Reply by mcrio** [reply]
I was checking the final document state in the DB and it was being updated, not the `dbSession.Advanced.HasChanges` property.

Someone from RavenDb team can comment why `HasChanges` and `HasChangesd()` do not account for metadata changes.
